src/tike/operators/cupy/cache.py

```diff
@@ -1,10 +1,17 @@
+__author__ = "Daniel Ching"
+__copyright__ = "Copyright (c) 2020, UChicago Argonne, LLC."
+
 import cupy as cp
 from cupyx.scipy.fft import fftn, ifftn
 from cupyx.scipy.fftpack import get_fft_plan
```

Comment:
Did you observe significant performance gain when using cached fft plans?

Reply:
:man_shrugging: I didn't actually test the performance impact for the Lamino operator, but I did observed performance gains for the Propagation operator in the past. It should be trivial to show the effect of caching however, let me run some tests from the profiling folder. Standby.

Reply:
## Without caches
```
35.210 template_algorithm  profile_lamino.py:36
└─ 35.210 reconstruct  tike/lamino/lamino.py:86
   ├─ 34.444 cgrad  tike/lamino/solvers/cgrad.py:8
   │  └─ 34.444 update_obj  tike/lamino/solvers/cgrad.py:21
   │     └─ 34.444 conjugate_gradient  tike/opt.py:90
   │        ├─ 23.393 line_search  tike/opt.py:16
   │        │  ├─ 13.489 cost_function  tike/lamino/solvers/cgrad.py:23
   │        │  │  └─ 13.486 cost  tike/operators/cupy/lamino.py:153
   │        │  │     └─ 13.108 fwd  tike/operators/cupy/lamino.py:65
   │        │  │        ├─ 10.784 eq2us  tike/operators/cupy/usfft.py:74
   │        │  │        │  ├─ 6.019 [self]  
   │        │  │        │  ├─ 2.543 checkerboard  tike/operators/cupy/usfft.py:245
   │        │  │        │  │  ├─ 1.386 _g  tike/operators/cupy/usfft.py:240
   │        │  │        │  │  │  ├─ 0.701 can_cast  <__array_function__ internals>:2
   │        │  │        │  │  │  │     [4 frames hidden]  <__array_function__ internals>, numpy
   │        │  │        │  │  │  └─ 0.685 [self]  
   │        │  │        │  │  └─ 0.502 [self]  
   │        │  │        │  ├─ 0.699 _fftn  tike/operators/cupy/lamino.py:15
   │        │  │        │  │  └─ 0.696 fftn  cupyx/scipy/fft/fft.py:169
   │        │  │        │  │        [86 frames hidden]  cupyx, cupy, <__array_function__ inte...
   │        │  │        │  ├─ 0.698 _get_kernel  tike/operators/cupy/usfft.py:11
   │        │  │        │  │  └─ 0.683 __getitem__  cupy/creation/ranges.py:244
   │        │  │        │  │        [20 frames hidden]  cupy, <__array_function__ internals>,...
   │        │  │        │  └─ 0.627 gather  tike/operators/cupy/lamino.py:68
   │        │  │        │     └─ 0.624 gather  tike/operators/cupy/lamino.py:136
   │        │  │        ├─ 1.879 checkerboard  tike/operators/cupy/usfft.py:245
   │        │  │        │  └─ 1.047 _g  tike/operators/cupy/usfft.py:240
   │        │  │        │     ├─ 0.619 can_cast  <__array_function__ internals>:2
   │        │  │        │     │     [4 frames hidden]  <__array_function__ internals>, numpy
   │        │  │        │     └─ 0.428 [self]  
   │        │  │        └─ 0.432 ifft2  cupyx/scipy/fft/fft.py:146
   │        │  │              [74 frames hidden]  cupyx, cupy, <__array_function__ inte...
   │        │  └─ 9.779 [self]  
   │        └─ 10.749 grad  tike/lamino/solvers/cgrad.py:26
   │           └─ 10.749 grad  tike/operators/cupy/lamino.py:157
   │              ├─ 7.603 adj  tike/operators/cupy/lamino.py:92
   │              │  ├─ 6.776 us2eq  tike/operators/cupy/usfft.py:175
   │              │  │  ├─ 4.941 [self]  
   │              │  │  ├─ 0.812 checkerboard  tike/operators/cupy/usfft.py:245
   │              │  │  │  └─ 0.394 _g  tike/operators/cupy/usfft.py:240
   │              │  │  ├─ 0.427 _fftn  tike/operators/cupy/lamino.py:15
   │              │  │  │  └─ 0.427 fftn  cupyx/scipy/fft/fft.py:169
   │              │  │  │        [32 frames hidden]  cupyx, cupy, <__array_function__ inte...
   │              │  │  └─ 0.384 scatter  tike/operators/cupy/lamino.py:95
   │              │  │     └─ 0.383 scatter  tike/operators/cupy/lamino.py:119
   │              │  └─ 0.540 checkerboard  tike/operators/cupy/usfft.py:245
   │              └─ 3.126 fwd  tike/operators/cupy/lamino.py:65
   │                 ├─ 2.274 eq2us  tike/operators/cupy/usfft.py:74
   │                 │  ├─ 0.960 checkerboard  tike/operators/cupy/usfft.py:245
   │                 │  │  └─ 0.474 _g  tike/operators/cupy/usfft.py:240
   │                 │  └─ 0.496 _fftn  tike/operators/cupy/lamino.py:15
   │                 │     └─ 0.496 fftn  cupyx/scipy/fft/fft.py:169
   │                 │           [59 frames hidden]  cupyx, cupy, <__array_function__ inte...
   │                 └─ 0.725 checkerboard  tike/operators/cupy/usfft.py:245
   │                    └─ 0.453 _g  tike/operators/cupy/usfft.py:240
   └─ 0.759 __init__  tike/operators/cupy/lamino.py:48
      └─ 0.759 _make_grids  tike/operators/cupy/lamino.py:161
         └─ 0.738 __getitem__  cupy/creation/ranges.py:244
               [21 frames hidden]  cupy
```
## With caches
```
31.810 template_algorithm  profile_lamino.py:36
└─ 31.810 reconstruct  tike/lamino/lamino.py:86
   ├─ 31.028 cgrad  tike/lamino/solvers/cgrad.py:8
   │  └─ 31.028 update_obj  tike/lamino/solvers/cgrad.py:21
   │     └─ 31.028 conjugate_gradient  tike/opt.py:90
   │        ├─ 20.477 line_search  tike/opt.py:16
   │        │  ├─ 11.989 cost_function  tike/lamino/solvers/cgrad.py:23
   │        │  │  └─ 11.985 cost  tike/operators/cupy/lamino.py:155
   │        │  │     └─ 11.882 fwd  tike/operators/cupy/lamino.py:61
   │        │  │        ├─ 10.093 eq2us  tike/operators/cupy/usfft.py:74
   │        │  │        │  ├─ 6.450 [self]  
   │        │  │        │  ├─ 2.274 checkerboard  tike/operators/cupy/usfft.py:245
   │        │  │        │  │  ├─ 1.287 _g  tike/operators/cupy/usfft.py:240
   │        │  │        │  │  │  ├─ 0.682 [self]  
   │        │  │        │  │  │  └─ 0.605 can_cast  <__array_function__ internals>:2
   │        │  │        │  │  │        [4 frames hidden]  <__array_function__ internals>, numpy
   │        │  │        │  │  ├─ 0.373 [self]  
   │        │  │        │  │  └─ 0.319 can_cast  <__array_function__ internals>:2
   │        │  │        │  │        [4 frames hidden]  <__array_function__ internals>, numpy
   │        │  │        │  ├─ 0.545 _get_kernel  tike/operators/cupy/usfft.py:11
   │        │  │        │  │  └─ 0.464 __getitem__  cupy/creation/ranges.py:244
   │        │  │        │  │        [24 frames hidden]  cupy, <__array_function__ internals>,...
   │        │  │        │  └─ 0.393 can_cast  <__array_function__ internals>:2
   │        │  │        │        [4 frames hidden]  <__array_function__ internals>, numpy
   │        │  │        └─ 1.671 checkerboard  tike/operators/cupy/usfft.py:245
   │        │  │           └─ 0.799 _g  tike/operators/cupy/usfft.py:240
   │        │  │              ├─ 0.410 can_cast  <__array_function__ internals>:2
   │        │  │              │     [4 frames hidden]  <__array_function__ internals>, numpy
   │        │  │              └─ 0.389 [self]  
   │        │  └─ 8.426 [self]  
   │        └─ 10.486 grad  tike/lamino/solvers/cgrad.py:26
   │           └─ 10.486 grad  tike/operators/cupy/lamino.py:159
   │              ├─ 7.635 adj  tike/operators/cupy/lamino.py:91
   │              │  ├─ 6.668 us2eq  tike/operators/cupy/usfft.py:175
   │              │  │  ├─ 5.271 [self]  
   │              │  │  ├─ 0.883 checkerboard  tike/operators/cupy/usfft.py:245
   │              │  │  │  └─ 0.443 _g  tike/operators/cupy/usfft.py:240
   │              │  │  └─ 0.331 scatter  tike/operators/cupy/lamino.py:94
   │              │  │     └─ 0.330 scatter  tike/operators/cupy/lamino.py:121
   │              │  └─ 0.767 checkerboard  tike/operators/cupy/usfft.py:245
   │              │     └─ 0.456 _g  tike/operators/cupy/usfft.py:240
   │              └─ 2.768 fwd  tike/operators/cupy/lamino.py:61
   │                 ├─ 1.994 eq2us  tike/operators/cupy/usfft.py:74
   │                 │  ├─ 0.892 checkerboard  tike/operators/cupy/usfft.py:245
   │                 │  │  └─ 0.542 _g  tike/operators/cupy/usfft.py:240
   │                 │  ├─ 0.411 fftn  tike/operators/cupy/lamino.py:67
   │                 │  │  └─ 0.409 _fftn  tike/operators/cupy/cache.py:43
   │                 │  └─ 0.335 _get_kernel  tike/operators/cupy/usfft.py:11
   │                 └─ 0.704 checkerboard  tike/operators/cupy/usfft.py:245
   │                    └─ 0.393 _g  tike/operators/cupy/usfft.py:240
   └─ 0.775 __init__  tike/operators/cupy/lamino.py:43
      └─ 0.775 _make_grids  tike/operators/cupy/lamino.py:163
         └─ 0.754 __getitem__  cupy/creation/ranges.py:244
               [19 frames hidden]  cupy
```
With caches is clearly faster using total time alone (I couldn't figure out the cuda profiler with CuPy so this is pyinstrument results). Each of the runs time varied around 1 second.

Reply:
I was able to get the CUDA profiler working. I didn't see any specific FFT plan functions, but I did notice that the line search was not searching the same number of point for cached vs uncached. I switched to a fixed step size so the number of function calls was the same, and the cached version was still slightly faster. I also noticed there is a lot of casting between 32 and 64 bit types going on, but the largest time consumer are the scatter gather methods from usfft.